src/nrnmpi/memory_usage.cpp

```diff
@@ -0,0 +1,51 @@
+/* do not want the redef in the dynamic load case */
+#include <nrnmpiuse.h>
+
+#if NRNMPI_DYNAMICLOAD
+#include <nrnmpi_dynam.h>
+#endif
+
+#include <nrnmpi.h>
+
+#if NRNMPI
+#include <mpi.h>
+#include <nrnmpidec.h>
+#include "nrnmpi_impl.h"
+
+#include <iostream>
+
+#include "neuron/container/memory_usage.hpp"
+
+static void sum_reduce_memory_usage(void* invec, void* inoutvec, int* len_, MPI_Datatype*) {
+    int len = *len_;
+
+    auto a = static_cast<neuron::container::MemoryUsage*>(invec);
+    auto ab = static_cast<neuron::container::MemoryUsage*>(inoutvec);
+
+    for (int i = 0; i < len; ++i) {
+        ab[i] += a[i];
+    }
+}
+
+void nrnmpi_memory_stats(neuron::container::MemoryStats& stats,
+                         neuron::container::MemoryUsage const& local_memory_usage) {
+    MPI_Op op;
+    MPI_Op_create(sum_reduce_memory_usage, /* commute = */ 1, &op);
+
+    MPI_Datatype memory_usage_mpitype;
+    MPI_Type_contiguous(sizeof(neuron::container::MemoryUsage), MPI_BYTE, &memory_usage_mpitype);
+    MPI_Type_commit(&memory_usage_mpitype);
+
+    MPI_Allreduce(
+        &local_memory_usage, &stats.total, 1, memory_usage_mpitype, op, nrnmpi_world_comm);
```

Comment:
From my glance at these changes and a naive interpretation of memory usage and how the user would actually use the ```pc.print_memory_stats()```, I'm confident that  the communicator should be ```nrnmpi_comm```. 

The only circumstance that this matters is when subworlds are used (otherwise ```nrnmpi_world_comm``` is the same as ```nrnmpi_comm```). When subworlds are in use, each subworld generally sets up, simulates, and tears down a complete model asynchronously with every other subworld.  In fact, if there are three distinct models I want to simulate and four subworlds, one of the subworlds will remain idle.  The practical use of subworlds is that there are either ```pc.nhost_world()``` subworlds or 1 subworld. In the former case,  ```pc.nhost() == 1``` and in the latter case ```pc.nhost() == pc.nhost_world()```.  Subworlds allow to simultaneously use the bulletin board and and network simulations involving global identifiers. The bulletin board use case is many simulations with different parameters. So if a subworld does a simulation, and the user is interested in memory, the user would add  memory usage printing to the  function executed by the subworld (i.e. setup, simulate, what is the memory usage,  teardown, now what is the memory usage).  Actually, this
suggests to me that ```pc.print_memory_stats``` should return a string which would be posted to the bulletin board and taken by the master to monitor potential memory leaks.

Reply:
From the slack discussion, I agree that we can keep the function as a printf.